### DISCUSSION TITLE
RocketMQ 5.x 消息粒度负载均衡

### DISCUSSION THREAD

**Discussion by dreamfly99** [apache/rocketmq#7826] [Q&A]
在新版本中有说明支持消息粒度负载均衡，包括普通队列、顺序队列，链接：https://rocketmq.apache.org/zh/docs/featureBehavior/08consumerloadbalance/
1. 经过测试，普通队列只要没有被当前消费者锁定的数据（即还未ACK的），都可以被其他消费者来提取
2. 但是在顺序队列中测试发现，消费者1对某个message group的消息，如果未提交ACK，这个队列中其他message group的消息也无法被其他消费者提取。

我自己的理解是，对顺序队列，应该是能对message group进行锁定，而不是对这个队列进行锁定。如果是给队列加锁，这岂不就是队列负载均衡了吗，若一个消费者hang住，也会导致整条队列的消息都得不到消费，跟消息粒度负载均衡定义的场景是不吻合的。

还望各位专家、小伙伴 给指点一二，感谢！

**Comment by Yelijah** [comment]
实测下来，顺序消息好像会在本地消费者执行失败一定次数后，其他消费者才能提取到

**Comment by lizhanhui** [comment]
> 我自己的理解是，对顺序队列，应该是能对message group进行锁定，而不是对这个队列进行锁定。如果是给队列加锁，这岂不就是队列负载均衡了吗，若一个消费者hang住，也会导致整条队列的消息都得不到消费，跟消息粒度负载均衡定义的场景是不吻合的。

可以开个Improvement Proposal，实现这个feature
